File: samples/java/DeliveryDispatch/Server/Configuration/src/main/java/systems/zlink/samples/deliverydispatch/server/configuration/SampleTopology.java
package systems.zlink.samples.deliverydispatch.server.configuration;

import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties("sample")
public record SampleTopology(
        String trackingChannelEndpoint,
        String trackingSpotEndpoint,
        String trackingSpotPubEndpoint,
        String customerStreamEndpoint,
        String courierStreamEndpoint,
        String dispatchHttpEndpoint,
        String dispatchSpotEndpoint,
        String dispatchChannelEndpoint,
        String customerSpotEndpoint,
        String customerSpotRouterEndpoint,
        String courierActorNode1SpotEndpoint,
        String courierActorNode2SpotEndpoint,
        String courierSessionSpotEndpoint,
        String redisEndpoint,
        String redisKeyPrefix,
        String courierNode,
        String logDirectory) {

    public SampleTopology {
        trackingChannelEndpoint = value(trackingChannelEndpoint, "tcp://127.0.0.1:48103");
        trackingSpotEndpoint = value(trackingSpotEndpoint, "tcp://127.0.0.1:48118");
        trackingSpotPubEndpoint = value(trackingSpotPubEndpoint, "tcp://127.0.0.1:48120");
        customerStreamEndpoint = value(customerStreamEndpoint, "tcp://127.0.0.1:48104");
        courierStreamEndpoint = value(courierStreamEndpoint, "tcp://127.0.0.1:48105");
        dispatchHttpEndpoint = value(dispatchHttpEndpoint, "http://127.0.0.1:48107");
        dispatchSpotEndpoint = value(dispatchSpotEndpoint, "tcp://127.0.0.1:48108");
        dispatchChannelEndpoint = value(dispatchChannelEndpoint, "tcp://127.0.0.1:48121");
        customerSpotEndpoint = value(customerSpotEndpoint, "tcp://127.0.0.1:48109");
        customerSpotRouterEndpoint = value(customerSpotRouterEndpoint, "tcp://127.0.0.1:48110");
        courierActorNode1SpotEndpoint =
                value(courierActorNode1SpotEndpoint, "tcp://127.0.0.1:48113");
        courierActorNode2SpotEndpoint =
                value(courierActorNode2SpotEndpoint, "tcp://127.0.0.1:48114");
        courierSessionSpotEndpoint = value(courierSessionSpotEndpoint, "tcp://127.0.0.1:48119");
        redisEndpoint = required(redisEndpoint, "redisEndpoint");
        redisKeyPrefix = value(redisKeyPrefix, "deliverydispatch:java:");
        courierNode = value(courierNode, "node1");
        logDirectory = required(logDirectory, "logDirectory");
    }

    public static String configPath(String[] args) {
        if (args.length != 2 || !"--config".equals(args[0]) || args[1].isBlank()) {
            throw new IllegalArgumentException("Usage: <role executable> --config <path>");
        }
        return args[1];
    }

    private static String value(String value, String fallback) {
        return value == null || value.isBlank() ? fallback : value;
    }

    private static String required(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("sample." + name + " is required");
        }
        return value;
    }
}

File: samples/kotlin/DeliveryDispatch/Server/Configuration/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/configuration/SampleTopology.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.configuration

import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties

data class SampleTopology(
    val registryUrl: String,
    val trackingUrl: String,
    val customerGatewayUrl: String,
    val courierSessionUrl: String,
    val courierSpotNode1Url: String,
    val courierSpotNode2Url: String,
    val dispatchUrl: String,
) {
    fun roleUrls(): Map<String, String> =
        mapOf(
            SampleNames.RegistryRole to registryUrl,
            SampleNames.TrackingRole to trackingUrl,
            SampleNames.CustomerGatewayRole to customerGatewayUrl,
            SampleNames.CourierSessionRole to courierSessionUrl,
            "${SampleNames.CourierSpotNodeRolePrefix}1" to courierSpotNode1Url,
            "${SampleNames.CourierSpotNodeRolePrefix}2" to courierSpotNode2Url,
            SampleNames.DispatchRole to dispatchUrl,
        )

    companion object {
        lateinit var RegistryPubEndpoint: String
        lateinit var RegistryRouterEndpoint: String
        lateinit var TrackingChannelEndpoint: String
        lateinit var TrackingSpotEndpoint: String
        lateinit var CustomerStreamEndpoint: String
        lateinit var CourierStreamEndpoint: String
        lateinit var DispatchHttpEndpoint: String
        lateinit var DispatchSpotEndpoint: String
        lateinit var DispatchChannelEndpoint: String
        lateinit var CustomerSpotEndpoint: String
        lateinit var CustomerSpotRouterEndpoint: String
        lateinit var CourierActorNode1SpotEndpoint: String
        lateinit var CourierActorNode2SpotEndpoint: String
        lateinit var CourierSessionSpotEndpoint: String
        lateinit var RedisEndpoint: String
        lateinit var RedisKeyPrefix: String
        lateinit var CourierNode: String
        lateinit var LogDirectory: String
        lateinit var StateDirectory: String

        fun configure(args: Array<String>) {
            require(args.size == 2 && args[0] == "--config" && args[1].isNotBlank()) {
                "Usage: <role> --config <path>"
            }
            val properties =
                Properties().also { values ->
                    Files.newBufferedReader(Path.of(args[1])).use(values::load)
                }
            RegistryPubEndpoint = value(properties, "registryPubEndpoint", "tcp://127.0.0.1:49101")
            RegistryRouterEndpoint =
                value(properties, "registryRouterEndpoint", "tcp://127.0.0.1:49102")
            TrackingChannelEndpoint =
                value(properties, "trackingChannelEndpoint", "tcp://127.0.0.1:49103")
            TrackingSpotEndpoint =
                value(properties, "trackingSpotEndpoint", "tcp://127.0.0.1:49118")
            CustomerStreamEndpoint =
                value(properties, "customerStreamEndpoint", "tcp://127.0.0.1:49104")
            CourierStreamEndpoint =
                value(properties, "courierStreamEndpoint", "tcp://127.0.0.1:49105")
            DispatchHttpEndpoint =
                value(properties, "dispatchHttpEndpoint", "http://127.0.0.1:49107")
            DispatchSpotEndpoint =
                value(properties, "dispatchSpotEndpoint", "tcp://127.0.0.1:49108")
            DispatchChannelEndpoint =
                value(properties, "dispatchChannelEndpoint", "tcp://127.0.0.1:49121")
            CustomerSpotEndpoint =
                value(properties, "customerSpotEndpoint", "tcp://127.0.0.1:49109")
            CustomerSpotRouterEndpoint =
                value(properties, "customerSpotRouterEndpoint", "tcp://127.0.0.1:49110")
            CourierActorNode1SpotEndpoint =
                value(properties, "courierActorNode1SpotEndpoint", "tcp://127.0.0.1:49113")
            CourierActorNode2SpotEndpoint =
                value(properties, "courierActorNode2SpotEndpoint", "tcp://127.0.0.1:49114")
            CourierSessionSpotEndpoint =
                value(properties, "courierSessionSpotEndpoint", "tcp://127.0.0.1:49119")
            RedisEndpoint = required(properties, "redisEndpoint")
            RedisKeyPrefix = value(properties, "redisKeyPrefix", "deliverydispatch:kotlin:")
            CourierNode = value(properties, "courierNode", "node1")
            LogDirectory = required(properties, "logDirectory")
            StateDirectory = required(properties, "stateDirectory")
        }

        private fun value(properties: Properties, name: String, fallback: String): String =
            properties.getProperty(name)?.takeIf(String::isNotBlank) ?: fallback

        private fun required(properties: Properties, name: String): String =
            requireNotNull(properties.getProperty(name)?.takeIf(String::isNotBlank)) {
                "Missing DeliveryDispatch sample config: $name"
            }
    }
}
